### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.demos.boring_classes`

### PR DESCRIPTION
**Pull request by krishnakalyan3**
Part of https://github.com/Lightning-AI/lightning/issues/13445

### CODE REVIEW

File: src/pytorch_lightning/demos/boring_classes.py [hunk 1]
```diff
@@ -26,7 +27,7 @@ def __init__(self, size: int, length: int):
         self.len = length
         self.data = torch.randn(length, size)
 
-    def __getitem__(self, index):
+    def __getitem__(self, index: int) -> Dict[str, Any]:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:30]:
```suggestion
    def __getitem__(self, index: int) -> Dict[str, torch.Tensor]:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 2]
```diff
@@ -40,7 +41,7 @@ def __init__(self, size: int, length: int):
         self.len = length
         self.data = torch.randn(length, size)
 
-    def __getitem__(self, index):
+    def __getitem__(self, index: int) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:44]:
```suggestion
    def __getitem__(self, index: int) -> torch.Tensor:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 3]
```diff
@@ -52,7 +53,7 @@ def __init__(self, size: int, count: int):
         self.count = count
         self.size = size
 
-    def __iter__(self):
+    def __iter__(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:56]:
```suggestion
    def __iter__(self) -> Generator[torch.Tensor, None, None]:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 4]
```diff
@@ -62,16 +63,16 @@ def __init__(self, size: int, count: int):
...
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:66]:
```suggestion
    def __iter__(self) -> Generator[torch.Tensor, None, None]:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 5]
```diff
@@ -90,14 +91,14 @@ def training_step(...):
         super().__init__()
         self.layer = torch.nn.Linear(32, 2)
 
-    def forward(self, x):
+    def forward(self, x: Any) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:94]:
```suggestion
    def forward(self, x: torch.Tensor) -> torch.Tenso9r:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 6]
```diff
@@ -90,14 +91,14 @@ def training_step(...):
...
         return self.layer(x)
 
-    def loss(self, batch, preds):
+    def loss(self, batch, preds) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:97]:
```suggestion
    def loss(self, batch, preds) -> torch.Tensor:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 7]
```diff
@@ -90,14 +91,14 @@ def training_step(...):
...
         # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
         return torch.nn.functional.mse_loss(preds, torch.ones_like(preds))
 
-    def step(self, x):
+    def step(self, x: Any) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:101]:
```suggestion
    def step(self, x: torch.Tensor) -> torch.Tensor:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 8]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
         loss = self.loss(batch, output)
         return {"loss": loss}
 
-    def training_step_end(self, training_step_outputs):
+    def training_step_end(self, training_step_outputs: Any) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:111]:
```suggestion
    def training_step_end(self, training_step_outputs: torch.Tensor) -> torch.Tensor:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 9]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         return training_step_outputs
 
-    def training_epoch_end(self, outputs) -> None:
+    def training_epoch_end(self, outputs: Any) -> None:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:114]: Can you please check the type here as well? Should be the same as in the LightningModule class

File: src/pytorch_lightning/demos/boring_classes.py [hunk 10]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         torch.stack([x["loss"] for x in outputs]).mean()
 
     def validation_step(self, batch, batch_idx):
         output = self(batch)
         loss = self.loss(batch, output)
         return {"x": loss}
 
-    def validation_epoch_end(self, outputs) -> None:
+    def validation_epoch_end(self, outputs: Any) -> None:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:122]: Please also check for LightningModule here

File: src/pytorch_lightning/demos/boring_classes.py [hunk 11]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         torch.stack([x["x"] for x in outputs]).mean()
 
     def test_step(self, batch, batch_idx):
         output = self(batch)
         loss = self.loss(batch, output)
         return {"y": loss}
 
-    def test_epoch_end(self, outputs) -> None:
+    def test_epoch_end(self, outputs: Any) -> None:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:130]: Also check for LM here

File: src/pytorch_lightning/demos/boring_classes.py [hunk 12]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         torch.stack([x["y"] for x in outputs]).mean()
 
-    def configure_optimizers(self):
+    def configure_optimizers(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:133]:
```suggestion
    def configure_optimizers(self) -> Tuple[List[torch.optim.Optimizer], List[torch.optim._LRScheduler]]:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 13]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
         lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
         return [optimizer], [lr_scheduler]
 
-    def train_dataloader(self):
+    def train_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:138]:
```suggestion
    def train_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 14]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         return DataLoader(RandomDataset(32, 64))
 
-    def val_dataloader(self):
+    def val_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:141]:
```suggestion
    def val_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 15]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         return DataLoader(RandomDataset(32, 64))
 
-    def test_dataloader(self):
+    def test_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:144]:
```suggestion
    def test_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 16]
```diff
@@ -107,43 +108,43 @@ def training_step(self, batch, batch_idx):
...
         return DataLoader(RandomDataset(32, 64))
 
-    def predict_dataloader(self):
+    def predict_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:147]:
```suggestion
    def predict_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 17]
```diff
@@ -168,21 +169,21 @@ def setup(self, stage: Optional[str] = None):
         if stage == "predict" or stage is None:
             self.random_predict = Subset(self.random_full, indices=range(64 * 3, 64 * 4))
 
-    def train_dataloader(self):
+    def train_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:172]:
```suggestion
    def train_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 18]
```diff
@@ -168,21 +169,21 @@ def setup(self, stage: Optional[str] = None):
...
         return DataLoader(self.random_train)
 
-    def val_dataloader(self):
+    def val_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:175]:
```suggestion
    def val_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 19]
```diff
@@ -168,21 +169,21 @@ def setup(self, stage: Optional[str] = None):
...
         return DataLoader(self.random_val)
 
-    def test_dataloader(self):
+    def test_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:178]:
```suggestion
    def test_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 20]
```diff
@@ -168,21 +169,21 @@ def setup(self, stage: Optional[str] = None):
...
         return DataLoader(self.random_test)
 
-    def predict_dataloader(self):
+    def predict_dataloader(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:181]:
```suggestion
    def predict_dataloader(self) -> torch.utils.DataLoader:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 21]
```diff
@@ -202,7 +203,7 @@ def __init__(self, out_dim: int = 10, learning_rate: float = 0.02):
         self.l1 = torch.nn.Linear(32, out_dim)
         self.learning_rate = learning_rate
 
-    def forward(self, x):
+    def forward(self, x: Any) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:206]:
```suggestion
    def forward(self, x: torch.Tensor) -> torch.Tensor:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 22]
```diff
@@ -211,12 +212,12 @@ def training_step(self, batch, batch_nb):
         loss = x.sum()
         return loss
 
-    def configure_optimizers(self):
+    def configure_optimizers(self) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:215]:
```suggestion
    def configure_optimizers(self) -> torch.optim.Optimizer:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 23]
```diff
@@ -225,7 +226,7 @@ def __init__(self):
         self.fc1 = nn.Linear(9216, 128)
         self.fc2 = nn.Linear(128, 10)
 
-    def forward(self, x):
+    def forward(self, x: Any) -> Tensor:
```
Comment (justusschock) [inline src/pytorch_lightning/demos/boring_classes.py:229]:
```suggestion
    def forward(self, x: torch.Tensor) -> Tensor:
```

**Review by justusschock — COMMENTED** [review]
Hey @krishnakalyan3 ,

thanks for the PR!

However, just slapping `Any` on every single type is not what we want to achieve here. We actually want meaningful types. I corrected a whole bunch of them. Would you mind resolving the others (and add missing imports)?

**Review by awaelchli — APPROVED** [review]
LGTM when Justus's comments addressed

**Comment by krishnakalyan3** [comment]
Thanks for the feedback @justusschock.

File: src/pytorch_lightning/demos/boring_classes.py [hunk 24]
```diff
@@ -90,60 +92,60 @@ def training_step(...):
...
         return torch.nn.functional.mse_loss(preds, torch.ones_like(preds))
 
-    def step(self, x):
+    def step(self, x: Tensor) -> Tensor:
         x = self(x)
         out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
         return out
 
-    def training_step(self, batch, batch_idx):
+    def training_step(self, batch: Tensor, preds: Tensor) -> Dict[str, Tensor]:  # type: ignore
```
Comment (carmocca) [inline src/pytorch_lightning/demos/boring_classes.py:107]: Where did you get `preds` from?

File: src/pytorch_lightning/demos/boring_classes.py [hunk 25]
```diff
@@ -90,60 +92,60 @@ def training_step(...):
...
 
-    def training_step_end(self, training_step_outputs):
+    def training_step_end(self, training_step_outputs: Any) -> Any:
         return training_step_outputs
 
-    def training_epoch_end(self, outputs) -> None:
+    def training_epoch_end(self, outputs: Any) -> None:
         torch.stack([x["loss"] for x in outputs]).mean()
 
     def validation_step(self, batch, batch_idx):
```
Comment (carmocca) [inline src/pytorch_lightning/demos/boring_classes.py:118]: Why isn't this annotated too?

File: src/pytorch_lightning/demos/boring_classes.py [hunk 26]
```diff
@@ -90,60 +92,60 @@ def training_step(...):
...
         output = self(batch)
         loss = self.loss(batch, output)
         return {"x": loss}
 
-    def validation_epoch_end(self, outputs) -> None:
+    def validation_epoch_end(self, outputs: Any) -> None:
         torch.stack([x["x"] for x in outputs]).mean()
 
     def test_step(self, batch, batch_idx):
```
Comment (carmocca) [inline src/pytorch_lightning/demos/boring_classes.py:126]: Same comment

File: src/pytorch_lightning/demos/boring_classes.py [hunk 27]
```diff
@@ -90,60 +93,60 @@ def training_step(...):
         super().__init__()
         self.layer = torch.nn.Linear(32, 2)
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Any:
```
Comment (rohitgr7) [inline src/pytorch_lightning/demos/boring_classes.py:96]:
```suggestion
    def forward(self, x: Tensor) -> Tensor:
```

File: src/pytorch_lightning/demos/boring_classes.py [hunk 28]
```diff
@@ -202,21 +205,21 @@ def __init__(self, out_dim: int = 10, learning_rate: float = 0.02):
         self.l1 = torch.nn.Linear(32, out_dim)
         self.learning_rate = learning_rate
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Any:
```
Comment (rohitgr7) [inline src/pytorch_lightning/demos/boring_classes.py:208]:
```suggestion
    def forward(self, x: Tensor) -> Tensor:
```

**Review by rohitgr7 — APPROVED** [review]

**Comment by krishnakalyan3** [comment]
@carmocca and @justusschock I have included the types from `LightningModule` modules. I still have mypy errors below.

I am not quite sure on what to do here.

```
src/pytorch_lightning/demos/boring_classes.py:96: error: Signature of "forward" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:96: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:96: note:          def forward(self, *args: Any, **kwargs: Any) -> Any
src/pytorch_lightning/demos/boring_classes.py:96: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:96: note:          def forward(self, x: Tensor) -> Any
src/pytorch_lightning/demos/boring_classes.py:108: error: Signature of "training_step" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:108: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:108: note:          def training_step(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:108: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:108: note:          def training_step(self, batch: Tensor, batch_idx: int) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:117: error: Invalid index type "str" for "Union[Tensor, Dict[str, Any]]"; expected type "Union[None, int, slice, Tensor, List[Any], Tuple[Any, ...]]"  [index]
src/pytorch_lightning/demos/boring_classes.py:119: error: Signature of "validation_step" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:119: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:119: note:          def validation_step(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any], None]
src/pytorch_lightning/demos/boring_classes.py:119: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:119: note:          def validation_step(self, batch: Tensor, batch_idx: int) -> Union[Tensor, Dict[str, Any], None]
src/pytorch_lightning/demos/boring_classes.py:125: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
src/pytorch_lightning/demos/boring_classes.py:125: note: Possible overload variants:
src/pytorch_lightning/demos/boring_classes.py:125: note:     def __getitem__(self, SupportsIndex) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:125: note:     def __getitem__(self, slice) -> List[Union[Tensor, Dict[str, Any]]]
src/pytorch_lightning/demos/boring_classes.py:125: error: Invalid index type "str" for "Union[Tensor, Dict[str, Any], List[Union[Tensor, Dict[str, Any]]]]"; expected type "Union[None, int, slice, Tensor, List[Any], Tuple[Any, ...]]"  [index]
src/pytorch_lightning/demos/boring_classes.py:127: error: Signature of "test_step" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:127: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:127: note:          def test_step(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any], None]
src/pytorch_lightning/demos/boring_classes.py:127: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:127: note:          def test_step(self, batch: Tensor, batch_idx: int) -> Union[Tensor, Dict[str, Any], None]
src/pytorch_lightning/demos/boring_classes.py:133: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
src/pytorch_lightning/demos/boring_classes.py:133: note: Possible overload variants:
src/pytorch_lightning/demos/boring_classes.py:133: note:     def __getitem__(self, SupportsIndex) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:133: note:     def __getitem__(self, slice) -> List[Union[Tensor, Dict[str, Any]]]
src/pytorch_lightning/demos/boring_classes.py:133: error: Invalid index type "str" for "Union[Tensor, Dict[str, Any], List[Union[Tensor, Dict[str, Any]]]]"; expected type "Union[None, int, slice, Tensor, List[Any], Tuple[Any, ...]]"  [index]
src/pytorch_lightning/demos/boring_classes.py:192: error: Signature of "training_step" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:192: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:192: note:          def training_step(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:192: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:192: note:          def training_step(self, batch: Tensor, batch_idx: int) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:196: error: Item "LightningOptimizer" of "Union[LightningOptimizer, List[LightningOptimizer]]" has no attribute "zero_grad"  [union-attr]
src/pytorch_lightning/demos/boring_classes.py:196: error: Item "List[LightningOptimizer]" of "Union[LightningOptimizer, List[LightningOptimizer]]" has no attribute "zero_grad"  [union-attr]
src/pytorch_lightning/demos/boring_classes.py:198: error: Item "List[LightningOptimizer]" of "Union[LightningOptimizer, List[LightningOptimizer]]" has no attribute "step"  [union-attr]
src/pytorch_lightning/demos/boring_classes.py:208: error: Signature of "forward" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:208: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:208: note:          def forward(self, *args: Any, **kwargs: Any) -> Any
src/pytorch_lightning/demos/boring_classes.py:208: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:208: note:          def forward(self, x: Tensor) -> Any
src/pytorch_lightning/demos/boring_classes.py:211: error: Signature of "training_step" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:211: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:211: note:          def training_step(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:211: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:211: note:          def training_step(self, batch: Tensor, batch_nb: int) -> Union[Tensor, Dict[str, Any]]
```

**Comment by carmocca** [comment]
The errors such as

```py
src/pytorch_lightning/demos/boring_classes.py:192: error: Signature of "training_step" incompatible with supertype "LightningModule"  [override]
src/pytorch_lightning/demos/boring_classes.py:192: note:      Superclass:
src/pytorch_lightning/demos/boring_classes.py:192: note:          def training_step(self, *args: Any, **kwargs: Any) -> Union[Tensor, Dict[str, Any]]
src/pytorch_lightning/demos/boring_classes.py:192: note:      Subclass:
src/pytorch_lightning/demos/boring_classes.py:192: note:          def training_step(self, batch: Tensor, batch_idx: int) -> Union[Tensor, Dict[str, Any]]
```
can be ignored as they will be gone in next mypy release, as they just closed: https://github.com/python/mypy/issues/5876

**Review by otaj — APPROVED** [review]

**Comment by krishnakalyan3** [comment]
Thanks, @otaj. I should have checked the asserts.

File: pyproject.toml [hunk 29]
```diff
@@ -51,7 +51,7 @@ warn_no_return = "False"
 module = [
     "pytorch_lightning.callbacks.progress.rich_progress",
     "pytorch_lightning.core.datamodule",
-    "pytorch_lightning.demos.boring_classes",
+    "pytorch_lightning.core.module",
```
Comment (carmocca) [inline pyproject.toml:54]: Why are you adding this file?

File: src/pytorch_lightning/demos/boring_classes.py [hunk 30]
```diff
@@ -168,26 +174,27 @@ def setup(self, stage: Optional[str] = None):
...
 class ManualOptimBoringModel(BoringModel):
-    def __init__(self):
+    def __init__(self) -> None:
         super().__init__()
         self.automatic_optimization = False
 
-    def training_step(self, batch, batch_idx):
+    def training_step(self, batch: Tensor, batch_idx: int) -> STEP_OUTPUT:  # type: ignore [override]
         opt = self.optimizers()
+        assert not isinstance(opt, list)
```
Comment (carmocca) [inline src/pytorch_lightning/demos/boring_classes.py:197]: It would be clearer to assert that this is an optimizer instance.

**Review by carmocca — APPROVED** [review]

File: pyproject.toml (hunk 29 above)
Comment (otaj) [reply pyproject.toml]: Sorry, that's my merge gonna wrong. Fixed now

File: src/pytorch_lightning/demos/boring_classes.py (hunk 30 above)
Comment (otaj) [reply src/pytorch_lightning/demos/boring_classes.py]: I agree, I just fixed it. Thanks!

**Pull request merged**